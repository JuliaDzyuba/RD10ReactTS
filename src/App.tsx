import React, { useEffect, useState } from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Route, Switch } from 'react-router-dom';
import { AppRoute, AppTheme } from './common/enum';
import { IData } from './common/types/data.type';
import Avatar from './components/Avatar';
import Button from './components/Button';
import Header from './components/Header';
import About from './pages/About';
import Experience from './pages/Experience';
import Form from './pages/Form';
import Home from './pages/Home';
import NotFound from './pages/NotFound';

const fetchData = async (url: string): Promise<IData[]> => {  
  const response = await fetch(url);
  const data = await response.json();
  return data;  
};

export const ThemeContext = React.createContext<{
  theme: string,
  toggleTheme: () => void,
}>({
  theme: AppTheme.LIGHT,
  toggleTheme() {},
});

const App: React.FC = () => {
  const [usersData, setUsersData] = useState<IData[]>([]);
  const [appTheme, setTheme] = useState<string>(AppTheme.LIGHT);
  
  useEffect(() => {
    (async () => {
      try {
        const data = await fetchData('./data.json');
        if(data.length) {
          setUsersData(data);
        }      
      } catch (error) {
        console.log(error);        
      }      
    })();
  }, []);

  const renderAvatar = () => <Avatar source={usersData[0]?.avatar} username={usersData[0]?.name}/>;

  const toggleTheme = () => {
    if(appTheme === AppTheme.LIGHT) {
      setTheme(AppTheme.DARK);
    } else {
      setTheme(AppTheme.LIGHT);
    }    
  };

  const addUser = () => {
    setUsersData([...usersData, {
      name: `new user ${Date.now()}`,
      currentRole: 'user',
      avatar: 'https://via.placeholder.com/300',
      summary: '',
      education: [],
      experience: [],
      skills: {
        mainStack: ['rr'],
        database: ['rr'],
        additionally: ['rr'],
        englishLevel: 'rr',
      },
    }])
  };

  return (
    <div className={`App ${appTheme}`}>
      <ThemeContext.Provider value={{
        theme: appTheme,
        toggleTheme: toggleTheme,
      }}>
        <Router>
          <Header />
          <Button onClick={addUser}/>
          <Switch>
            <Route exact path={AppRoute.ROOT} >
              <Home user={usersData[0]} renderAvatar={renderAvatar}/>
            </Route>
            <Route exact path={AppRoute.ABOUT}>
              <About user={usersData[0]}/>
            </Route>
            <Route exact path={AppRoute.EXPERIENCE}>
              <Experience user={usersData[0]}/>
            </Route>
            <Route exact path={AppRoute.FORM} >
              <Form users={usersData}/>
            </Route>
            <Route path={AppRoute.ANY} component={NotFound} />
          </Switch>
        </Router>
      </ThemeContext.Provider>    
    </div>
  );
};

export default App;
