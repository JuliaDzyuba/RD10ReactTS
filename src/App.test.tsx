import React from 'react';
import { render, screen } from '@testing-library/react';
import App from './App';

describe('App component', () => {
  test('should render', () => {
    const { container } = render(<App />);
    
    expect(container).toMatchSnapshot();
  });

  test('renders change theme button', () => {
    render(<App />);

    const changeBtn = screen.getByText(/change theme/i);

    expect(changeBtn).toBeInTheDocument();
  });
});
