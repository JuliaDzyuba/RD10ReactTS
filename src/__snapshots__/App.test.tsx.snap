// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render 1`] = `
<div>
  <div
    class="App light"
  >
    <header
      class="header"
    >
      <div
        class="container"
      >
        <nav
          class="nav"
        >
          <ul>
            <li>
              <a
                href="/"
              >
                Home
              </a>
            </li>
            <li>
              <a
                href="/about"
              >
                About
              </a>
            </li>
            <li>
              <a
                href="/experience"
              >
                Experience
              </a>
            </li>
            <li>
              <a
                href="/form"
              >
                Form
              </a>
            </li>
          </ul>
          <button
            type="button"
          >
            Change theme
          </button>
        </nav>
      </div>
    </header>
    <button
      type="button"
    >
      Add new user
    </button>
    <div
      class="container light"
    >
      <div
        class="homeWrapper"
      >
        <h2>
          Hi, I am 
        </h2>
        <div
          class="avatar"
        >
          <img />
        </div>
        <p
          class="role"
        />
      </div>
    </div>
  </div>
</div>
`;
